sql/codegen_alps.go

```diff
@@ -39,137 +40,63 @@ type alpsFiller struct {
 	X      string
 	Y      string
 
-	// Estimator
-	EstimatorCreatorCode string
-	TrainSpec            collection
-	EvalSpec             collection
+	// Train
+	ModelCreatorCode string
+	TrainClause      *resolvedTrainClause
 
-	// Config
-	OdpsConf    collection
-	DatasetConf collection
+	// ODPS
+	OdpsConf *gomaxcompute.Config
 }
 
-type collection map[string]string
-
-func (c collection) Get(key string, fallback string) string {
-	if v, ok := c[key]; ok {
-		return v
-	}
-	return fallback
-}
-
-func filter(attrs []*attribute, prefix string) []*attribute {
-	ret := make([]*attribute, 0)
+func modelCreatorCode(modelName string, attrs map[string]*attribute, args []string) (string, error) {
+	cl := make([]string, 0)
 	for _, a := range attrs {
-		if strings.EqualFold(a.Prefix, prefix) {
-			ret = append(ret, a)
-		}
-	}
-	return ret
-}
-
-func generateFeatureColumnCode(fcs []interface{}) (string, error) {
-	var codes = make([]string, 0, len(fcs))
-	for _, fc := range fcs {
-		if fc, ok := fc.(featureColumn); ok {
-			code, err := fc.GenerateCode()
-			if err != nil {
-				return "", nil
-			}
-			codes = append(codes, code)
-		} else {
-			return "", fmt.Errorf("input is not featureColumn interface")
-		}
-	}
-	return fmt.Sprintf("[%s]", strings.Join(codes, ",")), nil
-}
-
-func generateEstimatorCreator(estimator string, attrs []*attribute, args []string) (string, error) {
-	cl := make([]string, len(attrs))
-	for idx, a := range attrs {
 		code, err := a.GenerateCode()
 		if err != nil {
 			return "", err
 		}
-		cl[idx] = code
+		cl = append(cl, code)
 	}
 	if args != nil {
 		for _, arg := range args {
 			cl = append(cl, arg)
 		}
 	}
-	return fmt.Sprintf("tf.estimator.%s(%s)", estimator, strings.Join(cl, ",")), nil
+	//TODO(uuleon) support customized model
+	return fmt.Sprintf("tf.estimator.%s(%s)", modelName, strings.Join(cl, ",")), nil
 }
 
-func newALPSTrainFiller(pr *extendedSelect) (*alpsFiller, error) {
-	//TODO(uuleon): the scratchDir will be deleted after model uploading
-	scratchDir, err := ioutil.TempDir("/tmp", "alps_scratch_dir_")
+func newALPSTrainFiller(pr *extendedSelect, db *DB) (*alpsFiller, error) {
+	resolved, err := resolveTrainClause(&pr.trainClause)
 	if err != nil {
 		return nil, err
 	}
-	modelDir := fmt.Sprintf("%s/model/", scratchDir)
 
-	fcMap := map[string][]interface{}{}
-	fsMap := map[string]*featureSpec{}
+	csCode := make([]string, 0)
+	for _, css := range resolved.ColumnSpecs {
+		for _, cs := range css {
+			csCode = append(csCode, cs.ToString())
+		}
+	}
 
-	for target, columns := range pr.columns {
-		fcs, fss, err := resolveTrainColumns(&columns)
+	var odpsConfig = &gomaxcompute.Config{}
+	if db != nil {
```

Comment:
What happens if `db == nil`? Or, should we change to
```go
if db == nil {
        return nil, fmt.Errorf(...)
}
odpsConfig, err := gomaxcompute.ParseDSN(db.dataSourceName);
if err != nil {
        return nil, err
}
```

Reply:
`db == nil` happens in UnitTest

Reply:
@uuleon why do we need to test `db == nil`?